Commonalities/documentation/API-design-guidelines.md

```diff
@@ -1089,3 +1090,259 @@ following aspects:
+
+In order to ease developer adoption, the pattern for Resource-based event subscription should be consistent for all API providing this feature.
+
+4 operations must be defined:
+
+| operation | path | description |
+| ----- |	-----  |	 -----  | 
+| POST | `/event-subscriptions` |  Operation to request a event subscription.     |
+| GET | `/event-subscriptions` |  Operation to retrieve a list of event subscriptions - could be an empty list.  eg. `GET /event-subscriptions?type=ROAMING_STATUS&ExpireTime.lt=2023-03-17` |
+| GET | `/event-subscriptions/{eventSubscriptionsId}` | Operation to retrieve a event subscription |
+| DELETE | `/event-subscriptions/{eventSubscriptionsId}` | Operation to delete a event subscription |
+
+
+Note on the operation path:
+The recommended pattern is to use `/event-subscriptions` path for the subscription operation. But API design team, for specific case, has the option to append `/event-subscriptions` path with a prefix (e.g. `/roaming/event-subscriptions` and `/connectivity/event-subscriptions`). The rationale for using this alternate pattern should be explicitly provided (e.g. the notification source for each of the supported events may be completely different, in which case separating the implementations is beneficial). 
+
+Following table provides `/event-subscriptions` attributes
+
+| name | type | attribute description | cardinality |
+| ----- |	-----  |	 -----  |  -----  | 
+| notificationUrl | string | https callback address where the event notification must be POST-ed | mandatory |
+| notificationAuthToken | string | OAuth2 token to be used by the callback API endpoint. It MUST be indicated within HTTP Authorization header e.g. Authorization: Bearer $notificationAuthToken  | optional |
+| eventSubscriptionId | string | Identifier of the event subscription - This attribute must not be present in the POST request as it is provided by API server | mandatory in server response |
+| subscriptionExpireTime | string - datetime| Date when the event subscription should end. Provided by API requester. Server may reject the suscription if the period requested do not comply with Telco Operator policies (i.e. to avoid unlimited time subscriptions). In this case server returns exception 403 "SUBSCRIPTION_PERIOD_NOT_ALLOWED" | optional |
+| startsAt | string - datetime| Date when the event subscription will begin/began. This attribute must not be present in the `POST` request as it is provided by API server. It must be present in `GET` endpoints | optional |
+| expiresAt | string - datetime| Date when the event subscription will expire. This attribute must not be present in the `POST` request as it is provided by API server.  | optional |
+| correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |
```

Comment:
@PedroDiez  - Perhaps @sfnuser can elaborate because he asked for this. My understanding

- X-Correlator is used for request/response in header  - you pass it as consumer in an operation and you retrieve it in the response. The 'lifecycle' of this X-Correlator is the http request and it used for traceability.
- correlationId is provided by consumer in a subscription for event. For all coming events from this subscription, server has to value to correlationId.

I see 2 patterns with 2 distinct objectives but perhaps I missed something.

Reply:
@PedroDiez - As @bigludo7 mentioned, I made this proposal because earlier it was ambiguously mentioned that  `notificationAuthToken` can be used as an OAuth2 token or a plain correlation Id as per the API consumer's implementation. We agreed in the last commonalities call that `notificationAuthToken` to be of type OAuth2 token ONLY.

In that case, if there is a need for an API consumer to use plain correlation Id (and do not want to use OAuth2), they could use this field optional `correlationId` in the message `body`. However, I also agree X-Correlator-Id custom header can also be used for this purpose if agreed between the API consumer and the API provider. If we feel there is no use case for the `correlationId` field, I am OK for this field to be removed too.

Reply:
@bigludo7 @sfnuser @PedroDiez 
Happy for `correlationId` to be removed, but then how does the API provider know whether to include the `X-Correlator` header with notifications and what correlation value to use?

Is the rule that, if the API consumer includes an `X-Correlator` header when creating a subscription (i.e. with the `POST /event-subscriptions` call), then that header should not only be included in the response to that request, but also any notifications generated by that subscription? Or can the API consumer now not specify a specific correlation value to be used for notifications?

A brief description of expected use of the `X-Correlator` header in notifications would be useful here.

Reply:
@PedroDiez @bigludo7 - I prefer deleting these restrictions on `notificationAuthToken` because if I want to use a valid OAuth2 token, which does not meet these restrictions, my implementation is hampered without any reason. Hence I requested @PedroDiez [earlier](https://github.com/camaraproject/WorkingGroups/pull/173#discussion_r1188640321) to share the justifications on having these restrictions to better understand if I am missing anything.

Regarding `X-Correlation-Id`, I agree with @eric-murray. If there is an expectation on API provider to include the `X-Correlation-Id` (which is provided by API consumer earlier) in the notification messages, it needs to be explicitly stated as this is an `custom` agreement and not a standard.

Reply:
Understood the rationale @bigludo7 
@eric-murray @sfnuser When creating a subscription the identification of the subscription is provided: eventSubscriptionId

And that info can also be included within the events, so that correlation can be performed

Reply:
@PedroDiez
Indeed the API consumer will know the source of the notifications given the `eventSubscriptionId`, that was always true, but they cannot specify that value. So is your proposal that an `X-Correlator` header should not be included with notifications (or, at least, not using a correlation value specified by the API consumer)? That conflicts with your comment [here](https://github.com/camaraproject/WorkingGroups/pull/173#discussion_r1188594321).

Reply:
What I mean,

- my proposal is to use "x-correlator" in the context of every request/response.
- To correlate events we have the concept of eventSubscriptionId that is generated by API Server. All the events realted to a specific subscription can convey that `eventSubscriptionId`.

Reply:
I was scratching my head on this `x-correlator` reference and finally found it and its usage in the `Carrier Billing Checkout` API. So, for `instance` based event subscriptions, the respective API needs to provide `x-correlator` header value or something equivalent and it is upto the API (like Carrier Billing Checkout) to define this. Is this understanding correct?

Yes, for `resource` based event subscriptions, we have the `eventSubscriptionId`.

Reply:
Think we are mixing the two concepts:
- 'x-correlator' is to correlate a specific request context "request/response". It is a common header referenced in Commonalities. As commented by Ludovic already in this thread:

> X-Correlator is used for request/response in header - you pass it as consumer in an operation and you retrieve it in the response. The 'lifecycle' of this X-Correlator is the http request and it used for traceability.

- To have correlation between a subscriptionId and their related events -> eventSubscriptionId, as you mention